modules/netstat/netstat_linux.go

```diff
@@ -9,15 +9,152 @@ package netstat /* import "mig.ninja/mig/modules/netstat" */
+// include the host namespace in addition to other namespaces
+func procCandidates() (map[string][]string, error) {
+	var err error
+	ret := make(map[string][]string)
+
+	dirents, err := ioutil.ReadDir("/proc")
+	if err != nil {
+		return ret, err
+	}
+	for _, x := range dirents {
+		_, err = strconv.Atoi(x.Name())
+		if err != nil {
+			continue
+		}
+		nspath := path.Join("/proc", x.Name(), "ns", "net")
+		nsname, err := os.Readlink(nspath)
+		if err != nil {
+			// We don't treat this condition as fatal
+			continue
+		}
+		if _, ok := ret[nsname]; !ok {
+			ret[nsname] = make([]string, 0)
+		}
+		ret[nsname] = append(ret[nsname], path.Join("/proc", x.Name(), "net"))
+	}
+
+	return ret, nil
+}
+
+// Read lines from a network proc file; used by procNetNS()
+func procNetNSEntries(fname string) (ret []string, err error) {
+	fd, err := os.Open(fname)
+	if err != nil {
+		return ret, err
+	}
+	defer fd.Close()
+	scanner := bufio.NewScanner(fd)
+	scanner.Scan() // Skip the header
+	for scanner.Scan() {
+		if err := scanner.Err(); err != nil {
+			return ret, err
+		}
+		ret = append(ret, scanner.Text())
+	}
+	return ret, nil
+}
+
+// Given a network proc file (e.g., tcp would represent instances of
+// <pid>/net/tcp visible by the agent, return a list of proc network entries
+// for each active network namespace, including the host.
+//
+// This function works by building a cache of all current network namespaces
+// on the host based on /proc. From this list, we try to collect data from the
+// indicated file for at most one process per namespace. Since <pid>/net/*
+// represents network activity for the namespace the process is in, and not
+// the process itself we only need to be successful with one PID in each
+// namespace.
+func procNetNS(fname string) (ret []procNetLine, err error) {
+	defer func() {
+		if e := recover(); e != nil {
+			err = fmt.Errorf("procNetNS(): %v", e)
+		}
+	}()
+	procCache, err := procCandidates()
+	if err != nil {
+		panic(err)
+	}
+
+	// For each cached namespace, attempt to extract data from one
+	// process. We try until we find one to help work around the
+	// potential for the process to exit after we have cached the
+	// list.
+	for k, v := range procCache {
+		lns := make([]string, 0)
+		for _, x := range v {
+			lns, err = procNetNSEntries(path.Join(x, fname))
+			if err != nil {
+				// We don't treat this condition as fatal, just
+				// continue with the next candidate
+				continue
+			} else {
+				break
+			}
+		}
+		for _, x := range lns {
+			ret = append(ret, procNetLine{line: x, nsIdentifier: k})
+		}
+	}
+
+	return ret, nil
+}
+
+func procArpEntriesNS() (ret []procNetLine, err error) {
+	return procNetNS("arp")
+}
+
+func procArpEntries() (ret []procNetLine, err error) {
+	defer func() {
+		if e := recover(); e != nil {
+			err = fmt.Errorf("procArpEntries(): %v", e)
+		}
+	}()
+
+	// If we have namespace resolution enabled, use that
+	if namespaceMode {
+		return procArpEntriesNS()
+	}
```

Comment:
should namespace mode bypass reading `/proc/net/arp` entirely? Would reading both return duplicates?